apps/grader-host/src/index.js

```diff
@@ -457,311 +384,238 @@ function runJob(info, callback) {
+    try {
+      await container.wait();
+    } finally {
+      clearTimeout(timeoutId);
+    }
+
+    results.end_time = await timeReporter.reportEndTime(jobId);
+
+    const data = await container.inspect();
+    if (results.timedOut) {
+      logger.info('Container timed out');
+    } else {
+      logger.info(`Container exited with exit code ${data.State.ExitCode}`);
+    }
+    results.succeeded = !results.timedOut && data.State.ExitCode === 0;
+
+    await container.remove({
+      // Remove any volumes associated with this container
+      v: true,
+    });
+
+    // We made it through the Docker danger zone!
+    clearTimeout(jobTimeoutId ?? undefined);
+    logger.info('Reading course results');
+    // Now that the job has completed, let's extract the results
+    // First up: results.json
+    if (results.succeeded) {
+      await fs.readFile(path.join(tempDir, 'results', 'results.json')).then(
+        (data) => {
+          if (Buffer.byteLength(data) > 1024 * 1024) {
+            // Cap output at 1MB
+            results.succeeded = false;
+            results.message =
+              'The grading results were larger than 1MB. ' +
+              'If the problem persists, please contact course staff or a proctor.';
+            return;
           }
-          results.results = null;
-          callback(null);
-        }
-      },
-    ],
-    (err) => {
-      ERR(err, (err) => logger.error('runJob error:', err));
 
+          try {
+            const parsedResults = JSON.parse(data.toString());
+            results.results = sanitizeObject(parsedResults);
+            results.succeeded = true;
+          } catch (e) {
+            logger.error('Could not parse results.json:', e);
+            results.succeeded = false;
+            results.message = 'Could not parse the grading results.';
+          }
+        },
+        (err) => {
+          logger.error('Could not read results.json', err);
+          results.succeeded = false;
+          results.message = 'Could not read grading results.';
+        },
+      );
+    } else {
+      if (results.timedOut) {
+        results.message = `Your grading job did not complete within the time limit of ${timeout} seconds.\nPlease fix your code before submitting again.`;
+      }
+      results.results = null;
+    }
+  })()
+    .catch((err) => {
+      logger.error('runJob error', err);
+
+      results.succeeded = false;
+      results.message = err.toString();
+    })
+    .then(() => {
       // It's possible that we get here with an error prior to the global job timeout exceeding.
       // If that happens, Docker is still alive, but it just errored. We'll cancel
       // the timeout here if needed.
-      if (jobTimeoutId != null) {
-        clearTimeout(jobTimeoutId);
-      }
-
-      // If we somehow eventually get here after exceeding the global timeout,
-      // we should avoid calling the callback again
-      if (jobFailed) {
-        return;
-      }
```

Comment:
This is no longer necessary. We now use `Promise.race` to concurrently execute the timeout and the grading job, so if the timeout finishes first, it doesn't matter that the grading job promise resolves later; the result will just be ignored.